File: server/models/schedule.py
import enum

from sqlalchemy import Column, Enum, ForeignKey, Integer, String
from sqlalchemy.orm import relationship

from core.database import Base


class SessionType(str, enum.Enum):
    COURSE = "course"
    SEMINAR = "seminar"
    LAB = "lab"


class SessionStatus(str, enum.Enum):
    NORMAL = "normal"
    MOVED = "moved"
    CANCELED = "canceled"


class Schedule(Base):
    __tablename__ = "schedules"

    id = Column(Integer, primary_key=True, index=True)
    group_id = Column(Integer, ForeignKey("groups.id"), nullable=False)
    subject_id = Column(Integer, ForeignKey("subjects.id"), nullable=False)
    professor_id = Column(Integer, ForeignKey("professors.id"), nullable=False)
    room_id = Column(Integer, ForeignKey("rooms.id"), nullable=False)
    day = Column(String, nullable=False)  # Day of the week
    hour = Column(String, nullable=False)  # Time interval (e.g. 08:00–09:45)
    session_type = Column(Enum(SessionType), default=SessionType.COURSE, nullable=False)
    status = Column(Enum(SessionStatus), default=SessionStatus.NORMAL, nullable=False)
    notes = Column(String, nullable=True)
    version = Column(Integer, default=1, nullable=False)
    
    # Câmpuri pentru săptămâna impară (opționale)
    odd_week_subject_id = Column(Integer, ForeignKey("subjects.id"), nullable=True)
    odd_week_professor_id = Column(Integer, ForeignKey("professors.id"), nullable=True)
    odd_week_room_id = Column(Integer, ForeignKey("rooms.id"), nullable=True)

    group = relationship("Group")
    
    # Relații normale - trebuie să specificăm explicit foreign_keys pentru a evita ambiguitatea
    subject = relationship("Subject", foreign_keys=[subject_id])
    professor = relationship("Professor", foreign_keys=[professor_id])
    room = relationship("Room", foreign_keys=[room_id])
    
    # Relații pentru săptămâna impară
    odd_week_subject = relationship("Subject", foreign_keys=[odd_week_subject_id])
    odd_week_professor = relationship("Professor", foreign_keys=[odd_week_professor_id])
    odd_week_room = relationship("Room", foreign_keys=[odd_week_room_id])

File: client-student/types/schedule.ts
export type SessionType = 'course' | 'seminar' | 'lab';
export type SessionStatus = 'normal' | 'moved' | 'canceled';

export interface Group {
  id: number;
  code: string;
  year?: number | null;
  faculty?: string | null;
  specialization?: string | null;
}

export interface Professor {
  id: number;
  full_name: string;
  department?: string | null;
  email?: string | null;
}

export interface Subject {
  id: number;
  name: string;
  code: string;
  semester?: string | null;
}

export interface Room {
  id: number;
  code: string;
  building?: string | null;
  capacity?: number | null;
}

export interface Schedule {
  id: number;
  day: string;
  hour: string;
  session_type: SessionType;
  status: SessionStatus;
  notes?: string | null;
  version: number;
  group: Group;
  subject: Subject;
  professor: Professor;
  room: Room;
  odd_week_subject?: Subject | null;
  odd_week_professor?: Professor | null;
  odd_week_room?: Room | null;
  academic_year?: number | null;
  semester?: string | null;
  cycle_type?: string | null;
}

export interface ScheduleCreate {
  group_id: number;
  subject_id: number;
  professor_id: number;
  room_id: number;
  day: string;
  hour: string;
  session_type: SessionType;
  status: SessionStatus;
  notes?: string | null;
  odd_week_subject_id?: number | null;
  odd_week_professor_id?: number | null;
  odd_week_room_id?: number | null;
  academic_year?: number | null;
  semester?: string | null;
  cycle_type?: string | null;
}

export type ScheduleUpdate = Partial<ScheduleCreate>;

// Tipuri pentru evaluările periodice
export interface AssessmentSchedule {
  id: number;
  subject: string;
  groups_composition: string;
  professor_name: string;
  assessment_date: string;
  assessment_time: string;
  room_code: string;
  academic_year: number;
  semester: string;
  cycle_type?: string | null;
}

